tests/apps/test_similarity.py

```diff
@@ -222,142 +223,224 @@ def test_event_cardinality(photons, max_count, modes, expected):
+            ValueError, match="Loss parameter must take a value between zero and one"
+        ):
+            similarity.prob_orbit_exact(g, [1, 1, 1, 1], loss=2)
+
+    def test_prob_vacuum_orbit(self):
+        """Tests if the function gives the right probability for the empty orbit when the GBS
+        device has been configured to have zero mean photon number."""
+        graph = nx.complete_graph(5)
+        assert similarity.prob_orbit_exact(graph, [], 0) == 1.0
+
+    @pytest.mark.parametrize("k", [1, 3, 5, 7, 9])
+    def test_odd_photon_numbers(self, k):
+        """Test if function returns zero probability for odd number of total photons."""
+        graph = nx.complete_graph(10)
+        assert similarity.prob_orbit_exact(graph, [k]) == 0.0
+
+
+class TestProbEventExact:
+    """Tests for the function ``strawberryfields.apps.similarity.prob_event_exact``"""
+
+    def test_invalid_n_mean(self):
+        """Test if function raises a ``ValueError`` when the mean photon number is specified to
+        be negative."""
+        g = nx.complete_graph(5)
+        with pytest.raises(ValueError, match="Mean photon number must be non-negative"):
+            similarity.prob_event_exact(g, 2, 2, n_mean=-1)
+
+    def test_invalid_loss(self):
+        """Test if function raises a ``ValueError`` when the loss parameter is specified outside
+        of range."""
+        g = nx.complete_graph(5)
+        with pytest.raises(
+            ValueError, match="Loss parameter must take a value between zero and one"
+        ):
+            similarity.prob_event_exact(g, 2, 2, loss=2)
+
+    def test_invalid_photon_number(self):
+        """Test if function raises a ``ValueError`` when a photon number below zero is specified"""
+        g = nx.complete_graph(5)
+        with pytest.raises(ValueError, match="Photon number must not be below zero"):
+            similarity.prob_event_exact(g, -1, 2)
+
+    def test_low_count(self):
+        """Test if function raises a ``ValueError`` if ``max_count_per_mode`` is negative."""
+        g = nx.complete_graph(5)
+        with pytest.raises(
+            ValueError, match="Maximum number of photons per mode must be non-negative"
+        ):
+            similarity.prob_event_exact(g, 2, max_count_per_mode=-1)
+
+    def test_prob_vacuum_event(self):
+        """Tests if the function gives the right probability for an event with zero photons when
+        the GBS device has been configured to have zero mean photon number."""
+        graph = nx.complete_graph(5)
+        assert similarity.prob_event_exact(graph, 0, 0, 0) == 1.0
+
+    @pytest.mark.parametrize("k", [3, 5, 7, 9])
+    @pytest.mark.parametrize("nmax", [1, 2])
+    def test_odd_photon_numbers(self, k, nmax):
+        """Test if function returns zero probability for odd number of total photons."""
+        graph = nx.complete_graph(10)
+        assert similarity.prob_event_exact(graph, k, nmax) == 0.0
+
+
 class TestProbOrbitMC:
```

Comment:
Also needs to test that correct values are calculated for non-trivial specific cases